vault/request_handling.go

```diff
@@ -53,15 +54,30 @@ func (c *Core) HandleRequest(req *logical.Request) (resp *logical.Response, err
 	}
 
 	// We are wrapping if there is anything to wrap (not a nil response) and a
-	// TTL was specified for the token
-	wrapping := resp != nil && resp.WrapInfo != nil && resp.WrapInfo.TTL != 0
+	// TTL was specified for the token. Errors on a call should be returned to
+	// the caller, so wrapping is turned off if an error is hit and the error
+	// is logged to the audit log.
+	wrapping := resp != nil &&
+		err == nil &&
+		!resp.IsError() &&
+		resp.WrapInfo != nil &&
+		resp.WrapInfo.TTL != 0
 
 	if wrapping {
-		cubbyResp, err := c.wrapInCubbyhole(req, resp)
+		cubbyResp, cubbyErr := c.wrapInCubbyhole(req, resp)
 		// If not successful, returns either an error response from the
-		// cubbyhole backend or an error; if either is set, return
-		if cubbyResp != nil || err != nil {
-			return cubbyResp, err
+		// cubbyhole backend or an error; if either is set, set resp and err to
+		// those and continue so that that's what we audit log. Otherwise
+		// finish the wrapping and audit log that.
+		if cubbyResp != nil || cubbyErr != nil {
```

Comment:
No -- any value returned from the function is an error. `cubbyResp` will be an error response that came from the cubbyhole backend, whereas `cubbyErr` will be an error coming from the function itself. In either case, we want to trigger the error behavior.